[selfhost] [web] @colanode/web build failure

Hello!

I merged 0.3.0 into my fork and found a couple related issues with the web API client.
I ran the Docker build and this was the result:
```
PS C:\code\colanode> docker build -t jhyry9docks/hsf-colanode-web:v0.3.0-test --no-cache --build-arg VERSION=v0.3.0 --build-arg SHA=2199716ea1a523d55ab94ddc6a82683843c124af . -f apps\web\Dockerfile
[+] Building 511.3s (26/27)                                                                                                                                                                                                                                                                                                                                                             docker:desktop-linux 
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                                                                                                                                                    0.1s 
 => => transferring dockerfile: 1.29kB                                                                                                                                                                                                                                                                                                                                                                  0.1s 
 => [internal] load metadata for docker.io/library/nginx:1.27-alpine                                                                                                                                                                                                                                                                                                                                    2.2s 
 => [internal] load metadata for docker.io/library/node:22-alpine                                                                                                                                                                                                                                                                                                                                       2.2s 
 => [auth] library/node:pull token for registry-1.docker.io                                                                                                                                                                                                                                                                                                                                             0.0s 
 => [auth] library/nginx:pull token for registry-1.docker.io                                                                                                                                                                                                                                                                                                                                            0.0s 
 => [internal] load .dockerignore                                                                                                                                                                                                                                                                                                                                                                       0.1s 
 => => transferring context: 2B                                                                                                                                                                                                                                                                                                                                                                         0.1s 
 => [build  1/18] FROM docker.io/library/node:22-alpine@sha256:d2166de198f26e17e5a442f537754dd616ab069c47cc57b889310a717e0abbf9                                                                                                                                                                                                                                                                         0.3s 
 => => resolve docker.io/library/node:22-alpine@sha256:d2166de198f26e17e5a442f537754dd616ab069c47cc57b889310a717e0abbf9                                                                                                                                                                                                                                                                                 0.3s 
 => CACHED [stage-1 1/2] FROM docker.io/library/nginx:1.27-alpine@sha256:65645c7bb6a0661892a8b03b89d0743208a18dd2f3f17a54ef4b76fb8e2f2a10                                                                                                                                                                                                                                                               0.3s 
 => => resolve docker.io/library/nginx:1.27-alpine@sha256:65645c7bb6a0661892a8b03b89d0743208a18dd2f3f17a54ef4b76fb8e2f2a10                                                                                                                                                                                                                                                                              0.2s 
 => [internal] load build context                                                                                                                                                                                                                                                                                                                                                                      70.6s 
 => => transferring context: 82.72MB                                                                                                                                                                                                                                                                                                                                                                   70.4s 
 => CACHED [build  2/18] WORKDIR /app                                                                                                                                                                                                                                                                                                                                                                   0.0s 
 => [build  3/18] COPY package.json package-lock.json ./                                                                                                                                                                                                                                                                                                                                                0.5s 
 => [build  4/18] COPY ../../scripts scripts                                                                                                                                                                                                                                                                                                                                                            0.2s 
 => [build  5/18] COPY ../../assets assets                                                                                                                                                                                                                                                                                                                                                              0.7s
 => [build  6/18] COPY ../../packages/core/package.json packages/core/package.json                                                                                                                                                                                                                                                                                                                      0.6s
 => [build  7/18] COPY ../../packages/crdt/package.json packages/crdt/package.json                                                                                                                                                                                                                                                                                                                      0.6s
 => [build  8/18] COPY ../../packages/client/package.json packages/client/package.json                                                                                                                                                                                                                                                                                                                  0.3s
 => [build  9/18] COPY ../../packages/ui/package.json packages/ui/package.json                                                                                                                                                                                                                                                                                                                          0.4s
 => [build 10/18] COPY ../../apps/web/package.json apps/web/package.json                                                                                                                                                                                                                                                                                                                                0.3s
 => [build 11/18] RUN npm ci && npm i --save-dev lightningcss-linux-x64-musl                                                                                                                                                                                                                                                                                                                          171.2s
 => [build 12/18] COPY ../../packages/core packages/core                                                                                                                                                                                                                                                                                                                                                1.2s 
 => [build 13/18] COPY ../../packages/crdt packages/crdt                                                                                                                                                                                                                                                                                                                                                0.4s 
 => [build 14/18] COPY ../../packages/client packages/client                                                                                                                                                                                                                                                                                                                                            0.8s 
 => [build 15/18] COPY ../../packages/ui packages/ui                                                                                                                                                                                                                                                                                                                                                    0.9s 
 => [build 16/18] COPY ../../apps/web apps/web                                                                                                                                                                                                                                                                                                                                                          0.6s 
 => [build 17/18] COPY ../../tsconfig.base.json ./                                                                                                                                                                                                                                                                                                                                                      0.2s 
 => ERROR [build 18/18] RUN npm run build -w @colanode/core &&     npm run build -w @colanode/crdt &&     npm run build -w @colanode/client &&     npm run build -w @colanode/ui &&     npm run build -w @colanode/web &&     npm prune --production                                                                                                                                                  258.0s 
------
 > [build 18/18] RUN npm run build -w @colanode/core &&     npm run build -w @colanode/crdt &&     npm run build -w @colanode/client &&     npm run build -w @colanode/ui &&     npm run build -w @colanode/web &&     npm prune --production:
1.251
1.251 > @colanode/core@1.0.0 build
1.251 > tsc
1.251
20.70
20.70 > @colanode/crdt@1.0.0 build
20.70 > tsc
20.70
33.38
33.38 > @colanode/client@1.0.0 build
33.38 > tsc
33.38
73.99
73.99 > @colanode/ui@1.0.0 build
73.99 > tsc
73.99
156.3 
156.3 > @colanode/web@1.0.0 build
156.3 > vite build && tsc
156.3
157.2 vite v7.1.1 building for production...
158.9 transforming...
234.8 ✓ 4362 modules transformed.
237.2 rendering chunks...
239.6 computing gzip size...
239.8 dist/manifest.webmanifest                                     0.17 kB
239.8 dist/index.html                                               0.64 kB │ gzip:   0.33 kB
239.8 dist/assets/sqlite3-opfs-async-proxy-C_otN2ZJ.js              9.45 kB
239.8 dist/assets/sqlite3-worker1-bundler-friendly-DFv2tJFO.js    196.81 kB
239.8 dist/assets/sqlite3-TTJgSG-T.wasm                           856.32 kB │ gzip: 393.43 kB
239.8 dist/assets/dedicated-Qe9r58TO.js                           925.72 kB
239.8 dist/assets/index-C3BwQQdx.css                               87.87 kB │ gzip:  15.66 kB
239.8 dist/assets/workbox-window.prod.es5-CwtvwXb3.js               5.76 kB │ gzip:   2.37 kB
239.8 dist/assets/index-WmVkhGJC.js                             2,691.32 kB │ gzip: 801.22 kB
239.8
239.8 (!) Some chunks are larger than 500 kB after minification. Consider:
239.8 - Using dynamic import() to code-split the application
239.8 - Use build.rollupOptions.output.manualChunks to improve chunking: https://rollupjs.org/configuration-options/#output-manualchunks
239.8 - Adjust chunk size limit for this warning via build.chunkSizeWarningLimit.
239.8 ✓ built in 1m 22s
239.8 
239.8 PWA v1.0.3
239.8 Building src/workers/service.ts service worker ("es" format)...
242.5 vite v7.1.1 building for production...
242.6 transforming...
243.9 ✓ 69 modules transformed.
244.0 rendering chunks...
244.1 computing gzip size...
244.1 dist/service.mjs  87.87 kB │ gzip: 21.26 kB
244.1 ✓ built in 1.54s
244.3
244.3 PWA v1.0.3
244.3 mode      injectManifest
244.3 format:   es
244.3 precache  10 entries (4662.00 KiB)
244.3 files generated
244.3   dist/service.js
257.7 src/services/file-system.ts(143,26): error TS2769: No overload matches this call.
257.7   Overload 1 of 2, '(data: FileSystemWriteChunkType): Promise<void>', gave the following error.
257.7     Argument of type 'Uint8Array<ArrayBufferLike>' is not assignable to parameter of type 'FileSystemWriteChunkType'.
257.7       Type 'Uint8Array<ArrayBufferLike>' is not assignable to type 'ArrayBufferView<ArrayBuffer>'.
257.7         Types of property 'buffer' are incompatible.
257.7           Type 'ArrayBufferLike' is not assignable to type 'ArrayBuffer'.
257.7             Type 'SharedArrayBuffer' is not assignable to type 'ArrayBuffer'.
257.7               Types of property '[Symbol.toStringTag]' are incompatible.
257.7                 Type '"SharedArrayBuffer"' is not assignable to type '"ArrayBuffer"'.
257.7   Overload 2 of 2, '(data: FileSystemWriteChunkType): Promise<void>', gave the following error.
257.7     Argument of type 'Uint8Array<ArrayBufferLike>' is not assignable to parameter of type 'FileSystemWriteChunkType'.
257.7       Type 'Uint8Array<ArrayBufferLike>' is not assignable to type 'ArrayBufferView<ArrayBuffer>'.
257.7         Types of property 'buffer' are incompatible.
257.7           Type 'ArrayBufferLike' is not assignable to type 'ArrayBuffer'.
257.7             Type 'SharedArrayBuffer' is not assignable to type 'ArrayBuffer'.
257.7               Types of property '[Symbol.toStringTag]' are incompatible.
257.7                 Type '"SharedArrayBuffer"' is not assignable to type '"ArrayBuffer"'.
257.7 src/services/file-system.ts(189,26): error TS2769: No overload matches this call.
257.7   Overload 1 of 2, '(data: FileSystemWriteChunkType): Promise<void>', gave the following error.
257.7     Argument of type 'Uint8Array<ArrayBufferLike>' is not assignable to parameter of type 'FileSystemWriteChunkType'.
257.7       Type 'Uint8Array<ArrayBufferLike>' is not assignable to type 'ArrayBufferView<ArrayBuffer>'.
257.7         Types of property 'buffer' are incompatible.
257.7           Type 'ArrayBufferLike' is not assignable to type 'ArrayBuffer'.
257.7             Type 'SharedArrayBuffer' is not assignable to type 'ArrayBuffer'.
257.7               Types of property '[Symbol.toStringTag]' are incompatible.
257.7                 Type '"SharedArrayBuffer"' is not assignable to type '"ArrayBuffer"'.
257.7   Overload 2 of 2, '(data: FileSystemWriteChunkType): Promise<void>', gave the following error.
257.7     Argument of type 'Uint8Array<ArrayBufferLike>' is not assignable to parameter of type 'FileSystemWriteChunkType'.
257.7       Type 'Uint8Array<ArrayBufferLike>' is not assignable to type 'ArrayBufferView<ArrayBuffer>'.
257.7         Types of property 'buffer' are incompatible.
257.7           Type 'ArrayBufferLike' is not assignable to type 'ArrayBuffer'.
257.7             Type 'SharedArrayBuffer' is not assignable to type 'ArrayBuffer'.
257.7               Types of property '[Symbol.toStringTag]' are incompatible.
257.7                 Type '"SharedArrayBuffer"' is not assignable to type '"ArrayBuffer"'.
257.8 npm error Lifecycle script `build` failed with error:
257.8 npm error code 2
257.8 npm error path /app/apps/web
257.8 npm error workspace @colanode/web@1.0.0
257.8 npm error location /app/apps/web
257.8 npm error command failed
257.8 npm error command sh -c vite build && tsc
------
Dockerfile:33
--------------------
  32 |     # Build packages
  33 | >>> RUN npm run build -w @colanode/core && \
  34 | >>>     npm run build -w @colanode/crdt && \
  35 | >>>     npm run build -w @colanode/client && \
  36 | >>>     npm run build -w @colanode/ui && \
  37 | >>>     npm run build -w @colanode/web && \
  38 | >>>     npm prune --production
  39 |
--------------------
ERROR: failed to build: failed to solve: process "/bin/sh -c npm run build -w @colanode/core &&     npm run build -w @colanode/crdt &&     npm run build -w @colanode/client &&     npm run build -w @colanode/ui &&     npm run build -w @colanode/web &&     npm prune --production" did not complete successfully: exit code: 2
```

I investigated and found that `Uint8Array` is an `ArrayBufferLike` which is incompatible with the requirements of the acceptable parameter types for `FileSystemWritableFileStream.write(data)`.
Per [the MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemWritableFileStream/write#data) for the specified interface parameter:

> The file data to write, in the form of an [ArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer), [TypedArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray), [DataView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView), [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob), or string.

Which makes the error I'm getting make sense.
I'm not sure why this was not detected in the GH Actions builds.

Using a combination of [this](https://stackoverflow.com/questions/79345535/how-to-convert-bufferarraybufferlike-to-arraybuffer-in-typescript-5-7) and [this](https://typescript-eslint.io/rules/no-this-alias/) I was able to resolve the issue by ensuring the underlying type of the Uint8Array is an ArrayBuffer.

It appears as of the latest commit in `main` (fd8cfb63d5c1cbfaaa7058c1c4b007cc8b3fbbcb) that the code is still the same as the state in the tag commit for `v0.3.0` (2a9fea972fa37a98971c79b650214c8f8b231327).

I've created a fix and will create a PR for you to examine it.

--- a/apps/web/src/services/file-system.ts
+++ b/apps/web/src/services/file-system.ts
@@ -69,6 +69,20 @@ export class WebFileSystem implements FileSystem {
     return { parent, name: fileName };
   }
 
+  //Ensure the data type is compatible with the File Write API
+  private ensureArrayBuffer(data: Uint8Array): ArrayBuffer {
+    const arrayBuffer: ArrayBuffer =
+      data.buffer instanceof ArrayBuffer ?
+        data.buffer : new ArrayBuffer(data.byteLength);
+
+    if (!(data.buffer instanceof ArrayBuffer)) {
+      const view = new Uint8Array<ArrayBuffer>(arrayBuffer);
+      view.set(data);
+    }
+
+    return arrayBuffer;
+  }
+
   public async makeDirectory(path: string): Promise<void> {
     await this.getDirectoryHandle(path, true);
   }
@@ -138,9 +152,10 @@ export class WebFileSystem implements FileSystem {
     const file = await fileHandle.createWritable({ keepExistingData: false });
 
     return new WritableStream<Uint8Array>({
-      async write(chunk) {
-        // write each chunk directly to disk
-        await file.write(chunk);
+      write: async (chunk) => {
+        // Ensure data compatibility, write each chunk directly to disk
+        const arrayBuffer = this.ensureArrayBuffer(chunk);
+        await file.write(arrayBuffer);
       },
       async close() {
         await file.close(); // makes the data durable
@@ -184,9 +199,10 @@ export class WebFileSystem implements FileSystem {
     // Create or open the file
     const fileHandle = await parent.getFileHandle(name, { create: true });
 
-    // Create a writable stream and write the data
+    // Create a writable stream, ensure data compat, and write the data
     const writable = await fileHandle.createWritable();
-    await writable.write(data);
+    const arrayBuffer = this.ensureArrayBuffer(data);
+    await writable.write(arrayBuffer);
     await writable.close();
   }
 

